### PR TITLE
Step1

### PR DESCRIPTION
**Pull request by rwisecar**

### CODE REVIEW

File: src/server.py [hunk 1]
```diff
@@ -23,23 +24,44 @@ def server():
...
+            print(client_request)
+            conn.sendall(response_ok().encode('utf8'))
+            conn.sendall(client_request.encode('utf8'))
             conn, addr = server.accept()
 
         except KeyboardInterrupt:
             print('\nClosing echo server...')
             break
 
+        # except StandardError:
```
Comment (WillWeatherford) [inline src/server.py:41]: Make sure to remove any commented-out code.

File: src/server.py [hunk 2]
```diff
@@ -23,23 +24,44 @@ def server():
...
+    if version_info[0] == 2:
+        message = message.decode('utf-8')
+    message += u"\r\n"
+    return message
+
+
+def response_error():
+    """Send a HTTP response to client if bad request."""
+    message = 'HTTP/1.1 500 Internal Server Error\n'
+    if version_info[0] == 2:
```
Comment (WillWeatherford) [inline src/server.py:60]: No need to do a version check on this. A better approach to ensure all strings are unicode by using `from __future__ import unicode_literals`. This coerces all string literals to unicode in both versions. That means you can just explicitly and intentionaly encode to bytestring just once as you're sending the message through the connection.

File: src/test_servers.py [hunk 3]
```diff
@@ -1,8 +1,8 @@
 # encoding:utf-8
 """Test echo socket communication."""
 
-import pytest
-import sys
+# import pytest
+# import sys
 
 
 PARAMS_TABLE = [
```
Comment (WillWeatherford) [inline src/test_servers.py:8]: You definitely still need to test that your server and client can handle unicode non-ascii characters.

File: src/test_servers.py [hunk 4]
```diff
@@ -12,18 +12,30 @@
...
+#     """Test for unicode messages."""
+#     from client import client
+#     msg = 'CÅT'
+#     if sys.version_info[0] == 2:
+#         assert client(msg) == msg.decode('utf8')
+#     else:
+#         assert client(msg) == msg
+
+
+def test_response_ok():
```
Comment (WillWeatherford) [inline src/test_servers.py:32]: These tests are just testing the content of your response functions. You still need to assert that your server sends back the expected response.

File: src/server.py [hunk 5]
```diff
@@ -23,23 +24,44 @@ def server():
...
         try:
             buffer_length = 8
-            response = u''
-            while response[-2:] != u"\r\n":
+            client_request = u''
+            while client_request[-2:] != u"\r\n":
                 part = conn.recv(buffer_length)
-                response += part.decode('utf8')
-                print('Recieved: ', part)
+                client_request += part.decode('utf8')
```
Comment (WillWeatherford) [inline src/server.py:30]: Similar to with your echo server, you're going to have big problems handling unicode chars decoding them one chunk at a time. Also, multiple concatenations is a computationally inefficient way to build a string from chunks. A better approach is to append chunks to a list, then use the str.join method on the list.

File: src/server.py (hunk 2 above)
Comment (rwisecar) [reply src/server.py:60]: Hi Will, this is actually how Nick told us to manage unicode/ ascii errors, especially since import unicode_literals wasn't working. Claire and I would love to check in with you so we can better understand the problem- before the Monday deadline if at all possible.

File: src/server.py (hunk 5 above)
Comment (clair3st) [reply src/server.py:30]: Thanks Will. 
We removed the decoding for each 'fragment' of the incoming message from client. In terms of putting it in a list wouldn't we have to join for every new fragment to see if we reach our end marker? This would add more work? Not sure if I understand.
